cardano-node/src/Cardano/Node/Run.hs

```diff
@@ -219,8 +229,29 @@ handleSimpleNode p trace nodeTracers npm onKernel = do
     (isProducer nc)
     (customiseChainDbArgs $ dbValidation npm)
     identity -- No NodeParams customisation
-    $ \_registry nodeKernel -> onKernel nodeKernel
+    $ \registry nodeKernel -> do
+      case (stopOnSlotSynced npm, sfds) of
+        (MaxSlotNo maxSlot, InternalShutdown _r shutFd) -> do
+          traceWith (trTransformer TextualRepresentation MaximalVerbosity $
+                      severityNotice trace)
+            ("will terminate upon reaching " <> (pack $ show maxSlot) :: Text)
+          spawnSlotLimitTerminator registry (Node.getChainDB nodeKernel) maxSlot shutFd
+        _ -> pure ()
+      onKernel nodeKernel
  where
+  spawnSlotLimitTerminator
+    :: ResourceRegistry IO -> ChainDB.ChainDB IO blk -> SlotNo -> Fd -> IO ()
+  spawnSlotLimitTerminator registry chainDB maxSlot (Fd shutFd) =
+    void $ onEachChange registry identity Nothing
+      (getPoint <$> ChainDB.getTipPoint chainDB) $
+        \case
+          Origin -> pure ()
+          At pt -> when (blockPointSlot pt >= maxSlot) $ do
+            traceWith (trTransformer TextualRepresentation MaximalVerbosity $
+                       severityNotice trace)
+              ("Cardano.Node.Run.spawnSlotLimitTerminator: reached " <> show (blockPointSlot pt) <> ", ringing the shutdown doorbell" :: Text)
+            IO.hClose =<< IO.fdToHandle shutFd
```

Comment:
Can this be encapsulated in a function, e.g., `triggerShutdown`?

Reply:
Done, although it is unable to reuse the pattern match, so we now have to deal with "escaped" partiality.